prov/verbs/src/verbs_cq.c

```diff
@@ -230,47 +230,169 @@ static void fi_ibv_cq_read_data_entry(struct ibv_wc *wc, int i, void *buf)
 		wc->byte_len : 0;
 }
 
-static int fi_ibv_match_ep_id(struct slist_entry *entry, const void *ep_id)
+static int fi_ibv_match_ep_id(struct slist_entry *entry,
+			      const void *ep_id)
 {
-	struct fi_ibv_msg_epe *epe = container_of(entry, struct fi_ibv_msg_epe, entry);
+	struct fi_ibv_msg_epe *epe =
+		container_of(entry, struct fi_ibv_msg_epe, entry);
 
-	if (epe->ep->ep_id == (uint64_t) ep_id)
-		return 1;
-
-	return 0;
+	return (epe->ep->ep_id == (uint64_t)ep_id);
 }
 
-/* Must call with cq->lock held */
-ssize_t fi_ibv_poll_cq(struct fi_ibv_cq *cq, struct ibv_wc *wc)
+static inline int fi_ibv_handle_internal_signal_wc(struct fi_ibv_cq *cq,
+						   struct ibv_wc *wc)
 {
 	struct fi_ibv_msg_epe *epe;
 	struct slist_entry *entry;
-	ssize_t ret;
 
-	ret = ibv_poll_cq(cq->cq, 1, wc);
-	if (ret <= 0)
-		return ret;
-
-	if (wc->opcode == IBV_WC_RECV || wc->opcode == IBV_WC_RECV_RDMA_WITH_IMM)
-		return ret;
-
-	/* TODO Handle the case when app posts a send with same wr_id */
-	if ((wc->wr_id & cq->wr_id_mask) != cq->send_signal_wr_id)
-		return ret;
-
-	entry = slist_remove_first_match(&cq->ep_list, fi_ibv_match_ep_id, (void *)wc->wr_id);
+	entry = slist_remove_first_match(&cq->ep_list,
+					 fi_ibv_match_ep_id,
+					 (void *)wc->wr_id);
 	if (!entry) {
 		VERBS_WARN(FI_LOG_CQ, "No matching EP for :"
 			   "given signaled send completion\n");
 		return -FI_EOTHER;
 	}
 	epe = container_of(entry, struct fi_ibv_msg_epe, entry);
 	ofi_atomic_sub32(&epe->ep->unsignaled_send_cnt,
-			VERBS_SEND_SIGNAL_THRESH(epe->ep));
+			 VERBS_SEND_SIGNAL_THRESH(epe->ep));
 	ofi_atomic_dec32(&epe->ep->comp_pending);
 	util_buf_release(cq->epe_pool, epe);
 
-	return 0;
+	return FI_SUCCESS;
+}
+
+static inline int fi_ibv_poll_outstanding_cq(struct fi_ibv_msg_ep *ep,
+					     struct fi_ibv_cq *cq)
+{
+	struct fi_ibv_wre *wre;
+	struct fi_ibv_wce *wce;
+	struct fi_ibv_msg_ep *tmp_ep;
+	struct ibv_wc wc;
+	ssize_t ret;
+
+	ret = ibv_poll_cq(cq->cq, 1, &wc);
+	if (ret <= 0)
+		return ret;
+
+	if ((wc.opcode == IBV_WC_RECV) ||
+	    (wc.opcode == IBV_WC_RECV_RDMA_WITH_IMM) ||
+	    ((wc.wr_id & cq->wr_id_mask) != cq->send_signal_wr_id)) {
+		wre = (struct fi_ibv_wre *)(uintptr_t)wc.wr_id;
+		if (wre->ep) {
+			if ((wre->ep != ep) &&
+			    (wc.status != IBV_WC_WR_FLUSH_ERR)) {
+				wce = util_buf_alloc(cq->wce_pool);
+				if (!wce)
+					return -FI_EAGAIN;
+				memset(wce, 0, sizeof(*wce));
+				wc.wr_id = (uintptr_t)wre->context;
+				wce->wc = wc;
+				slist_insert_tail(&wce->entry, &cq->wcq);
+			}
+		} else {
+			/* if `ep == NULL` it means that WRE
+			 * has already been released */
+			return ret;
+		}
+		tmp_ep = wre->ep;
+		wre->ep = NULL;
+		util_buf_release(tmp_ep->wre_pool, wre);
```

Comment:
We also need to remove wre from tmp_ep->wre_list.

Reply:
Ops, It was missed during previous applying of comments (it was on 282 line). Thanks for your attention!